CoAP support

arturkow2000 opened the pull request.

arturkow2000 committed:
wip CoAP support

arturkow2000 committed:
pal_nrf: implement heap support

arturkow2000 committed:
Add basic CoAP client

arturkow2000 committed:
pal_nrf: implement monotonic timer, reduce usage of unsafe code

This commit implements get_time_ms() which returns time from boot in
milliseconds. Additionally unsafe code usage in timer driver has been
reduced.

arturkow2000 committed:
coap: pass response from server to callback

arturkow2000 committed:
coap: rename src/coap.rs to src/coap/mod.rs

Let's be consistent with other modules.

arturkow2000 committed:
coap: remove unused code

arturkow2000 committed:
coap: support for timeouts while waiting for response

arturkow2000 committed:
coap: relax lifetime bounds on callback

This is required to allow using closures which capture non-static
elements, like self - if closure captures it cannot outlive the object
it captures, hence cannot be 'static.

arturkow2000 committed:
Begin work on FobnailClient class.

FobnailClient is responsible for communication with server running on
attester. It is layered on top of CoapClient.

arturkow2000 committed:
coap: begin working on support for non-reliable transfers and
piggybacked responses.

arturkow2000 committed:
use build.sh to pass RUSTFLAGS

Using .cargo/config file breaks vscode rust-analyzer, instead use
build.sh to pass RUSTFLAGS when building for nRF target.

arturkow2000 committed:
smoltcp: increase CoAP buffer size, use heap allocated buffers

Increase CoAP buffer size to 1500 (from 512) so we don't loose large
packets. Since we are now using heap, we can use it to allocate network
buffers, doing so reduces code complexity as we don't have to keep
thousands of variables for each socket we create.

arturkow2000 committed:
coap: support piggybacked responses

arturkow2000 committed:
Update FobnailClient

arturkow2000 committed:
FobnailClient: add timeout to Idle state.

On error other than timeout sleep 5 seconds before retrying. Doing so
prevents flooding console.

arturkow2000 committed:
coap: support request re-sending in CoapClient

This command brings support for re-sending request in case of timeout
while waiting for response. FobnailClient does not need to handle
timeouts itself anymore. Also this fixes a problem when the same request
when re-sent has a new token each time. This can be mishandled by server
as a separate request (instead of being ignored as duplicate).

arturkow2000 committed:
coap: don't print warning when receiving response with unknown token

Doing so causes Fobnail to print warning messages on receiving
duplicate packets.

arturkow2000 committed:
coap: implement message IDs

Message IDs are used by CoAP protocol for deduplication purposes and are
required for proper operation.

arturkow2000 committed:
client: don't handle timeouts during init phase

Timeouts are now handled by CoapClient and no longer need special
treatment during init phase, since packet is automatically retransmitted
as long as retry_count > 0.

arturkow2000 committed:
client: print response from server as string

krystian-hebel commented on src/main.rs:
```diff
@@ -10,18 +10,66 @@ extern crate pal_pc as pal;
+// need.
+const COAP_CLIENT_BUF_LEN: usize = 1500;
+
+// FIXME: should not use hardcoded server addresses, instead server should be
+// discovered
```
Actually, Fobnail should make host use this (or any other in this network) address so it won't have to be manually assigned by user. Maybe hardcoded response to DHCP query from host?

krystian-hebel replied:
Right, by hardcoded I meant just IP, without having to implement a database, assignment algorithm etc. on DHCP server's side.

Ack on the updated comment.

krystian-hebel reviewed (approved):
I haven't looked too deep into the code but compiled application seems to work so I'm giving weak approve.

arturkow2000 committed:
Update comments.

Clarify TIMER3 usage, update comment regarding static IP addresses -
attester IP address should be assigned by DHCP server in Fobnail
firmware.

arturkow2000 commented on pal/pal_nrf/src/lib.rs:
```diff
@@ -81,7 +84,10 @@ pub fn init() {
     );
 
     // configure TIMER2 to be used for delays
```
I updated comment:
```
// configure TIMER2 to be used for delays
// configure TIMER3 as a freerunning monotonic counter
```
